trussworks/react-uswds
# fix: Fix TimePicker filter behavior. Misc ComboBox Fixes

# Summary

The MVP TimePicker filtering behavior was identical to our ComboBox implementation, since it was just wrapping the ComboBox.

USWDS made some changes to filtering to the ComboBox (see #1108). This PR also brings those enhancements to our reactified library. 

There was also some existing behavior of our combobox that did not match the USWDS combobox (whether that was always the case, or due to the USWDS changes since our original implementation, I don't know, but I fixed as many as I found here along the way). 

Previous ComboBox behavior can be tested at our hosted storybook: https://trussworks.github.io/react-uswds/?path=/story/components-form-controls-combo-box--default-combo-box-with-prop-options

Current USWDS ComboBox behavior: https://designsystem.digital.gov/components/combo-box/

Previous TimePicker behavior can be tested at our hosted storybook: https://trussworks.github.io/react-uswds/?path=/story/components-form-controls-time-picker--complete-time-picker

Current USWDS TimePicker behavior: https://designsystem.digital.gov/components/time-picker/
 
The following issues were found & resolved (and tests added to hopefully cement the fixes since they're not all obvious):

> **Note:** The videos show the following:
>
> | USWDS Model Behavior | Current Behavior | Fixed Behavior (this PR) |
> | --- | --- | --- |

- Pressing enter on the input without an exact match, but with a previously selected option should revert to the selected option.

https://user-images.githubusercontent.com/15805554/115453292-78be7180-a1ed-11eb-87ac-756e8369caa8.mov

- Typing into the input to filter options, then Blurring without selecting should reset the filter (previously it would keep the filter until the input value was changed)

https://user-images.githubusercontent.com/15805554/115453475-aefbf100-a1ed-11eb-85b8-fe58fbfe7d62.mov

- Selecting an option should give that option the selected style in the dropdown, even if the input filter allows more than one option to be seen.

https://user-images.githubusercontent.com/15805554/115453435-a3102f00-a1ed-11eb-8ef7-649ba5d09686.mov

- When no option is selected, but the dropdown is open, the first option should have the focused style applied (but not actually be focused)

https://user-images.githubusercontent.com/15805554/115453664-dfdc2600-a1ed-11eb-86bf-ef2bb4af3a05.mov

- When the ComboBox is opened and scrolled, then closed, and reopened it should either be at the top, OR if an item is selected, that item should be scrolled into view.

https://user-images.githubusercontent.com/15805554/115453699-e79bca80-a1ed-11eb-949c-16827cd3d027.mov

- When an option is selected, but the filter is changed such that the selected option is not visible, the first partial match should be highlighted with the focused style (but not actually be focused)

https://user-images.githubusercontent.com/15805554/115453740-f5e9e680-a1ed-11eb-8ba6-dd92cb318afe.mov

- TimePicker should not filter via hiding options, but should move the focus styled element to the first match (and scroll to said match)

https://user-images.githubusercontent.com/15805554/115453777-fda98b00-a1ed-11eb-8b82-ab1ecccdf3d6.mov

- TimePicker should pass custom filtering logic into combobox to generate regex to find options. This allows `5:3p` when typed to match `5:30pm`:
![image](https://user-images.githubusercontent.com/15805554/114793630-c6a02900-9d58-11eb-9a9a-ab74fac8fdb3.png)

https://user-images.githubusercontent.com/15805554/115453829-0ef29780-a1ee-11eb-8751-bad4290dc92d.mov


## Related Issues or PRs

closes: #1108

## How To Test

The unit tests I added help to cover the scenarios I fixed.

To compare to our existing storybook or the USWDS implementation, use the deployed storybook or checkout this branch and run storybook locally.

@christopherhuii I would love your help in testing this in a project since I want to make sure I didn't break the blurring behavior and focus trap fixes you made recently (I did some testing locally, but I'm not sure I covered everything)
